CSUB-684: Support ECDSA PKs in CLI

pLabarta opened the pull request:
# Description of proposed changes

<describe what this PR is about and why we want it>

---
Practical tips for PR review & merge:

- [ ] All GitHub Actions report PASS
- [ ] Newly added code/functions have unit tests
  - [ ] Coverage tools report all newly added lines as covered
  - [ ] The positive scenario is exercised
  - [ ] Negative scenarios are exercised, e.g. assert on all possible errors
  - [ ] Assert on events triggered if applicable
  - [ ] Assert on changes made to storage if applicable
- [ ] Modified behavior/functions - try to make sure above test items are covered
- [ ] Integration tests are added if applicable/needed

pLabarta committed:
add: replace getSeed with initKeyring funcs that support ecdsa

pLabarta committed:
implement initKeyring in bond

pLabarta committed:
implement initKeyring in show-address

pLabarta committed:
implement initKeyring in chill

pLabarta committed:
implement initKeyring in distribute-rewards

pLabarta committed:
implement initKeyring in send

pLabarta committed:
implement initKeyring in set-keys

pLabarta committed:
make --ecdsa a global flag

pLabarta committed:
implement initKeyring in unbond

pLabarta committed:
implement initKeyring in validate

pLabarta committed:
implementinitKeyring in withdraw

pLabarta committed:
implement initKeyring in wizard

pLabarta committed:
remove unused imports

pLabarta committed:
yarn format

pLabarta committed:
yarn format #2

pLabarta committed:
add ecdsa tests to integration-tests

Commit:
Auto-update creditcoin-js type definitions

pLabarta committed:
add CC_PK to camelCase exceptions

pLabarta committed:
remove unused import

pLabarta committed:
fix: was using older prettier version

atodorov commented on scripts/cc-cli/src/utils/account.ts:
```diff
@@ -14,69 +15,83 @@ export function initECDSAKeyringPairFromPK(pk: string) {
+): Promise<KeyringPair> {
+  const interactive = options.input;
+  const ecdsa = options.ecdsa;
+  envVar = ecdsa ? "CC_PK" : envVar;
```
If I understand this correctly then when `ecdsa = true` the only ENV variable this function will read from is `CC_PK` overriding whatever was passed-in as argument. 

In the case of `wizard` this means both Stash & Controller accounts will be one and the same account, correct ?

atodorov replied:
A number of comments & questions:

0) Passing an argument to a function and then overriding it internally is very confusing IMO.
1) Are we sure we want both Stash & Controller accounts to be the same if ecdsa is true ? 
2) Do we want to keep using the `CC_PK` ENV variable or still use the `CC_STASH_SEED` / `CC_CONTROLLER_SEED` variables but instead of passing seeds, pass the primary key? (Potentially we can rename them to allow for the overriden meaning)

pLabarta replied:
Did not think about how confusing it is now, I agree. Keeping the same ENV variable as passed seems the best and I would rename them to `SECRET`.

atodorov commented on scripts/cc-cli/src/utils/account.ts:
```diff
@@ -14,69 +15,83 @@ export function initECDSAKeyringPairFromPK(pk: string) {
-        message: `Specify a seed phrase for the ${accountRole} account`,
-        validate: (seed) => mnemonicValidate(seed),
+        message: `Specify a ${inputName} for the ${accountRole} account`,
+        validate: (input) => validateInput(input),
```
During interactive mode we can still use 2 different Stash & Controller accounts.

atodorov commented on scripts/cc-cli/src/utils/account.ts:
```diff
@@ -14,69 +15,83 @@ export function initECDSAKeyringPairFromPK(pk: string) {
+  const ecdsa = options.ecdsa;
+  envVar = ecdsa ? "CC_PK" : envVar;
+  const inputName = ecdsa ? "private key" : "seed phrase";
+  const validateInput = ecdsa ? () => true : mnemonicValidate;
```
No input validation when ecdsa = true. 

Q: Are there functions to validate a private key  ? We could at least validate the length .

atodorov replied:
How do I create / generate a key in this ecdsa format ?

atodorov commented on scripts/cc-cli/src/test/integration-tests/useECDSA.test.ts:
```diff
@@ -0,0 +1,51 @@
+    );
+  });
+
+  it("should be able to send CTC", async () => {
```
IMO split this into a `send.test.ts` file and parametrize the test function to execute once with `--ecdsa` and once without. 

Test parametrization should be supported natively by our current version of Jest, see Option #2 here:
https://medium.com/@suman.maity112/parameterized-tests-with-jest-2932f42b1509

atodorov replied:
If that works, then we need to exercise wizard & validator tests using `--ecdsa` too because these are the most common execution paths.

atodorov reviewed (changes requested):
Added some comments and questions.

pLabarta committed:
keep envVar consistent when using ecdsa

pLabarta committed:
add: ecdsa key validation

pLabarta committed:
update and split tests

pLabarta committed:
format *sigh*

atodorov commented on scripts/cc-cli/src/test/integration-tests/validatorCycle.test.ts:
```diff
@@ -259,7 +259,7 @@ describe("integration test: validator manual setup", () => {
+          CC_CONTROLLER_SECRET: controllerSeed,
         },
       },
     );
```
Needs testing scenarios/parameters for `--ecdsa`.

atodorov commented on scripts/cc-cli/src/test/integration-tests/wizard.test.ts:
```diff
@@ -15,8 +15,8 @@ describe("integration test: validator wizard setup", () => {
-          CC_STASH_SEED: stash.seed,
-          CC_CONTROLLER_SEED: controller.seed,
+          CC_STASH_SECRET: stash.seed,
+          CC_CONTROLLER_SECRET: controller.seed,
```
Needs testing scenarios/parameters for `--ecdsa`.

atodorov commented on scripts/cc-cli/src/test/integration-tests/showAddress.test.ts:
```diff
@@ -0,0 +1,45 @@
+import execa from "execa";
+import { fundAddressesFromSudo } from "./helpers";
+import { parseAmountInternal } from "../../utils/parsing";
```
I see lots of warnings for unused imports. For some reason `eslint` is exiting with 0 here.

atodorov replied:
@pLabarta can you cherry-pick 

https://github.com/gluwa/creditcoin/pull/1259/commits/b17660d8875fcd59d1b5d46f3c2982f47ea57a49

and 

https://github.com/gluwa/creditcoin/pull/1259/commits/9bd8029a9b6c6cb0929cba8b526e7dabca37bab1.


Then it will leave only the warnings here:
```
/home/runner/work/creditcoin/creditcoin/scripts/cc-cli/src/utils/account.ts
Warning:   77:23  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
Warning:   78:30  warning  Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion
```

Commit:
Auto-update creditcoin-js type definitions

pLabarta committed:
remove unused imports and vars

pLabarta committed:
add: randomTestAccount helper

pLabarta committed:
upgrade wizard test to cover ecdsa/phrase

pLabarta committed:
upgrade showAddress test to use randomTestAccount

pLabarta committed:
use randomTestAccount in send test

pLabarta committed:
export validateECDSAKey

pLabarta committed:
upgrade validatorCycle test to cover ecdsa

pLabarta committed:
fix: formatting, finally

pLabarta committed:
fix: bad params in prompt command

atodorov committed:
Treat eslint warnings as errors & exit 1

things like unused imports are warnings and easlint will exit with 0
so we don't get to notice them!  This change will cause CI jobs to fail
so that we can notice these issues and fix them!

atodorov committed:
Remove unused imports

pLabarta committed:
appease linter

atodorov commented on scripts/cc-cli/src/utils/account.ts:
```diff
@@ -14,69 +15,105 @@ export function initECDSAKeyringPairFromPK(pk: string) {
+    if (validateInput(input)) {
+      return generateKeyring(input);
     } else {
       throw new Error(
```
Integration tests fail with:

```
FAIL src/test/integration-tests/wizard.test.ts (68.479 s)
  ● integration test: validator wizard setup › new validator should appear as waiting after running using an ecdsa pk

    Command failed with exit code 1: creditcoin-cli wizard --amount 1000 --url ws://localhost:9945 --ecdsa
    Error: Seed phrase provided in environment variable CC_STASH_SECRET is invalid.
    🧙 Running staking wizard...

      31 |
      32 |       // Run wizard setup with 1k ctc ang to pair with node Bob
    > 33 |       execa.commandSync(
         |             ^
      34 |         `creditcoin-cli wizard --amount 1000 --url ${BOB_NODE_URL} ${
      35 |           ecdsa ? "--ecdsa" : ""
      36 |         }`,
```

nathanwhit commented on scripts/cc-cli/src/utils/account.ts:
```diff
@@ -14,69 +15,105 @@ export function initECDSAKeyringPairFromPK(pk: string) {
+  throw new Error(`Error: Could not retrieve ${inputName}`);
+}
+
+export function validateECDSAKey(pk: string): boolean {
+  const keyring = initECDSAKeyringPairFromPK(pk);
+  const msg = "";
+  const sig = keyring.sign(msg);
+  return keyring.verify(msg, sig, keyring.publicKey);
+}
```
I don't see how this can ever return `false`. Any given string is a valid seed for a keypair, so `initECDSAKeyringPairFromPK` will always return a valid keypair (just potentially with an unexpected private key if you typoed the PK). Then this function is effectively testing that a valid keypair creates a valid signature, which will always be the case (unless polkadotJS is broken).

pLabarta replied:
Yes, previous validation was just `() => true`. All strings are valid ECDSA PKs, so there is no check. We can go back to `true` in the future if we want to prevent wasting resources on creating a keyring and simulating a signature.

nathanwhit commented on scripts/cc-cli/src/utils/account.ts:
```diff
@@ -14,69 +15,105 @@ export function initECDSAKeyringPairFromPK(pk: string) {
+      throw new Error(
+        "Error: Unexpected type; could not retrieve seed phrase or PK from environment variable.",
+      );
+  }
```
```suggestion
  if (envVar === undefined) {
    throw new Error(
      "Error: Unexpected type; could not retrieve seed phrase or PK from environment variable.",
    );
  }
  return envVar;
```

nathanwhit reviewed (dismissed):
A couple of small comments but nothing blocking

pLabarta committed:
replace switch with if

Co-authored-by: Nathan Whitaker <17734409+nathanwhit@users.noreply.github.com>
